### DISCUSSION TITLE
reducing code duplication for different types with similar code

### DISCUSSION THREAD

**Discussion by henriksjostrom** [prisma/prisma#24335] [Q&A] [answered]
### Question

In typescript I have a function that looks roughly like this but with around 30 cases most of them are the same.

```ts
async function deleteFeature(
  db: DB,
  type: "Feature1" | "Feature2" | "Feature3",
  id: number
): void {
  switch (type) {
    case 'Feature1':
      await db.feature1.delete({
        where: { externalId: id },
      })
    case 'Feature2':
      await db.feature2.delete({
        where: { externalId: id },
      })
    case 'Feature3':
      await db.feature3.delete({
        where: { externalId: id },
        include: specialcase,
      })
  }
}
```

Is there any resonable way to convert it to be more like

```ts
async function deleteFeature(
  db: DB,
  type: "Feature1" | "Feature2" | "Feature3",
  id: number
): void {
  switch (type) {
    case 'Feature3':
      await db.feature3.delete({
        where: { externalId: id },
        include: specialcase,
      })
    default:
      await db[type].delete({
        where: { externalId: id },
      })
  }
}
```
So that I can avoid duplicating the code for every single added type when they are almost the same?

The greatest challenge seems to be the camelCase vs PascalCase of the enum and the generated instance names from the db connection in TypeMap meta.modelProps.

### Expected behavior (optional)

Some way of avoiding listing every single type every time would be very nice and massively reduce the bloat in my codebase as I have several functions that looks roughly like this

- OS: Ubuntu
- Database: PostgreSQL
- Node.js version: v20.14.0

- Run `prisma -v` to see your Prisma version and paste it: 5.14.0

**Comment by prkagrawal** [comment]
Just convert pascal to camel case:

```javascript
const specialcase = {}; // Define your special case as needed

// Helper function to convert PascalCase to camelCase
const toCamelCase = (str: string) => str.charAt(0).toLowerCase() + str.slice(1);

async function deleteFeature(
  db: DB,
  type: "Feature1" | "Feature2" | "Feature3",
  id: number
): Promise<void> {
  const camelCaseType = toCamelCase(type);
  
  if (type === 'Feature3') {
    await db.feature3.delete({
      where: { externalId: id },
      include: specialcase,
    });
  } else {
    // @ts-ignore
    await db[camelCaseType].delete({
      where: { externalId: id },
    });
  }
}

//
// Example usage
//
const db: DB = {
  feature1: {
    delete: async ({ where: { externalId } }) => {
      console.log(`Deleted feature1 with id: ${externalId}`);
    },
  },
  feature2: {
    delete: async ({ where: { externalId } }) => {
      console.log(`Deleted feature2 with id: ${externalId}`);
    },
  },
  feature3: {
    delete: async ({ where: { externalId }, include }) => {
      console.log(`Deleted feature3 with id: ${externalId} and special case: ${include}`);
    },
  },
};

(async () => {
  await deleteFeature(db, 'Feature1', 1);
  await deleteFeature(db, 'Feature2', 2);
  await deleteFeature(db, 'Feature3', 3);
})();
```

**Comment by henriksjostrom** [comment]
How do I preserve type safety while doing that? With the ts ignore all additional features that would be special cases are now runtime errors which would be highly problematic. Instead the mapping has to happen at compile time somehow

**Reply by millsp** [reply]
There isn't an easy way to preserve type safety and you will need to cast as `any`, as @prkagrawal suggested. This is because you are trying to generalize many signatures into a single one. Due to the nature of how generics work in TypeScript, you will be better off closing an eye on type-safety.

**Comment by prkagrawal** [comment]
Just change `await db[camelCaseType].delete({` to `await (db as any)[camelCaseType].delete({` and you can remove ts-ignore

**Comment by henriksjostrom** [comment, marked as answer]
I think I managed to solve it, the main challenge seems to have been that all the queries were of a different type but since only part of the query that is the same is used it could be handled by a slightly narrower typing.

```ts

// Helper function to convert PascalCase to camelCase
function uncapitalize<T extends string>(str: T) {
  return (str.charAt(0).toLowerCase() + str.slice(1)) as Uncapitalize<T>
}
async function deleteFeature(
  db: DB,
  type: "Feature1" | "Feature2" | "Feature3",
  id: number
): Promise<void> {  
  if (type === 'Feature3') {
    await db.feature3.delete({
      where: { externalId: id },
      include: specialcase,
    });
  } else {
    const query = {
        where: { externalId: id },
      }
    const deleter: (del: typeof query) => Promise<void> =
        db[uncapitalize(type)].delete
    await deleter(query);
  }
}
```
